src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -3491,27 +3492,51 @@ private void EmitConstantExpression(TypeSymbol type, ConstantValue constantValue
         private bool TryEmitStringLiteralAsUtf8Encoded(ConstantValue constantValue, SyntaxNode syntaxNode)
         {
             // Emit long strings into data section so they don't overflow the UserString heap.
-            if (constantValue.IsString &&
-                constantValue.StringValue.Length > _module.Compilation.DataSectionStringLiteralThreshold)
+            if (!constantValue.IsString)
             {
-                if (Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Text_Encoding__get_UTF8, _diagnostics, syntax: syntaxNode) == null |
-                    Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Text_Encoding__GetString, _diagnostics, syntax: syntaxNode) == null)
-                {
-                    return false;
-                }
+                return false;
+            }
 
-                Cci.IFieldReference field = _module.TryGetOrCreateFieldForStringValue(constantValue.StringValue, syntaxNode, _diagnostics.DiagnosticBag);
-                if (field == null)
+            bool utf8Required = _module.PreviousGeneration != null && _module.PreviousGeneration.UserStringStreamLength > EmitBaseline.UserStringHeapSizeLimit;
+            if (!utf8Required)
+            {
+                var threshold = _module.Compilation.DataSectionStringLiteralThreshold;
+                if (threshold == null || constantValue.StringValue.Length <= threshold)
                 {
                     return false;
                 }
+            }
 
+            var field = tryGetOrCreateField();
+            if (field != null)
+            {
                 _builder.EmitOpCode(ILOpCode.Ldsfld);
                 _builder.EmitToken(field, syntaxNode, _diagnostics.DiagnosticBag);
                 return true;
             }
 
+            if (utf8Required)
+            {
+                _diagnostics.Add(ErrorCode.ERR_TooManyUserStrings_RestartRequired, syntaxNode);
```

Comment:
This error won't be reported if the previous generation was under the limit but the current emit goes over the limit, right? (What will happen in that case, is that tested?) So it's only reported in the (I imagine not very common) scenario where the previous generation goes just over the limit with the last string?

Reply:
Good catch. I think we'll need to keep track of the total size of string literals being emitted and switch automatically to data section strings when exceeded.

Reply:
So, we have a bigger problem. There are constructs that emit string constants and do not go via this code path (e.g. string switch).
We'll need to move this logic lower.

Reply:
Good catch, I missed that during implementation of the feature

Reply:
Good catch, I missed that during implementation of the feature

Reply:
Fixed